### PR TITLE
Fixing Dollar Sign Escape

### PR DESCRIPTION
**Pull request by Algorithm5838**
Improve the implementation of the dollar sign escape to address the issues with LaTeX.
Update: Done and ready for merge.

### CODE REVIEW

**Comment by daiaji** [comment]
https://github.com/quantizor/markdown-to-jsx/issues/485#issuecomment-1873346921
I wonder if it would be better to switch to markdown-to-jsx, perhaps to make the code more lightweight. The markdown-to-jsx example recommends some code to directly define the scope in which LaTeX rendering takes effect, which looks more elegant than the symbolic substitution used today.

**Comment by Algorithm5838** [comment]
Unfortunately, I am unfamiliar with it, maybe others can help out. I just fixed the inconsistent escaping.

File: app/components/markdown.tsx [hunk 1]
```diff
@@ -99,21 +99,19 @@ export function PreCode(props: { children: any }) {
...
+      isInCodeBlock = !isInCodeBlock;
+      return line;
     }
-
-    escapedText += char;
-  }
-
-  return escapedText;
+    if (!isInCodeBlock) {
+      return line.replace(/(?<!`.*|\\)\$\d+([,.](\d+[,.])?\d+)?(?!.*\$\B)(?!.*`)/g, '\\$&');
```
Comment (H0llyW00dzZ) [inline app/components/markdown.tsx:110]: you may need to handle inside single backtick, table as well

**Review by H0llyW00dzZ — COMMENTED** [review]
@Algorithm5838 Just letting you know

File: app/components/markdown.tsx (hunk 1 above)
Comment (Algorithm5838) [reply app/components/markdown.tsx:110]: Thank you for your input. It does handle inside single backticks, unless one used a backtick in one line and another in the next line. Also, it does handle tables as well.

Comment (Algorithm5838) [reply app/components/markdown.tsx:110]: How about this?
```
function escapeDollarNumber(text: string): string {
  let isInCode = false;
  return text.split('\n').map(line => {
    if (line.trim() === '```' || line.trim() === '`') {
      isInCode = !isInCode;
      return line;
    }
    if (!isInCode) {
      return line.replace(/(?<!`.*|\\)\$\d+([,.](\d+[,.])?\d+)?(?!.*\$\B)(?!.*`)/g, '\\$&');
    } else {
      return line;
    }
  }).join('\n');
}
```

Comment (H0llyW00dzZ) [reply app/components/markdown.tsx:110]: > Thank you for your input. It does handle inside single backticks, unless one used a backtick in one line and another in the next line. Also, it does handle tables as well.

However, it still doesn't handle properly in your [`deployment`](https://4363-vercelpreviewdeployment.pr.nextchat.dev/)

image

Comment (Algorithm5838) [reply app/components/markdown.tsx:110]: I see. I will try to fix it.
Thank you.

Comment (Algorithm5838) [reply app/components/markdown.tsx]: Fixed.

File: app/components/markdown.tsx [hunk 2]
```diff
@@ -100,14 +100,20 @@ export function PreCode(props: { children: any }) {
...
+  let isInMultilineCode = false;
   return text.split('\n').map(line => {
     if (line.trim() === '```' || line.trim() === '`') {
-      isInCode = !isInCode;
+      isInMultilineCode = !isInMultilineCode;
       return line;
     }
-    if (!isInCode) {
-      return line.replace(/(?<!\\)\$\d+([,.](\d+[,.])?\d+)?(?!.*\$\B)(?!`+)/g, '\\$&');
+    if (!isInMultilineCode) {
```
Comment (H0llyW00dzZ) [inline app/components/markdown.tsx:109]: it's better

image


I think it should be fine for inside a table

**Review by H0llyW00dzZ — COMMENTED** [review]
@Algorithm5838

File: app/components/markdown.tsx (hunk 2 above)
Comment (Algorithm5838) [reply app/components/markdown.tsx]: I just made another edits. The previous one had issues rendering latex. Check the new one. Hopefully there are no more issues.

File: app/components/markdown.tsx [hunk 3]
```diff
@@ -100,20 +100,20 @@ export function PreCode(props: { children: any }) {
...
+      isInBlockCode = !isInBlockCode;
       return line;
     }
-    if (!isInMultilineCode) {
-      return line.replace(/(`[^`]*`)|(\$[0-9]+(?:[,.][0-9]+)*)/g, (match, inlineCode, number) => {
-        if (inlineCode) {
-          return inlineCode;
+    if (!isInBlockCode) {
+      return line.split(/(`.*?`)/g).map((segment, index) => {
+        if (index % 2 === 0) {
```
Comment (H0llyW00dzZ) [inline app/components/markdown.tsx:111]: now it only handle correctly for left but not the right one

image

**Review by H0llyW00dzZ — COMMENTED** [review]
@Algorithm5838

**Comment by daiaji** [comment]
😄

File: app/components/markdown.tsx [hunk 4]
```diff
@@ -109,7 +109,7 @@ function escapeDollarNumber(text: string): string {
     if (!isInBlockCode) {
       return line.split(/(`.*?`)/g).map((segment, index) => {
         if (index % 2 === 0) {
-          return segment.replace(/(?<!\\)\$\d+([,.](\d+[,.])?\d+)?(?!.*\$\B)/g, '\\$&');
+          return segment.replace(/(?<!\\)\$\d+([,.](\d+[,.])?\d+)?(?!.*\$\.)/g, '\\$&');
```
Comment (H0llyW00dzZ) [inline app/components/markdown.tsx:112]: better

image

**Review by H0llyW00dzZ — COMMENTED** [review]
@Algorithm5838

File: app/components/markdown.tsx (hunk 4 above)
Comment (Algorithm5838) [reply app/components/markdown.tsx:112]: Unfortunately, there is a new issue with: $1$, or any other number. It always escapes numbers. I will look into it.

Comment (H0llyW00dzZ) [reply app/components/markdown.tsx:112]: > Unfortunately, there is a new issue with: `$1$`, or any other number. It always escapes numbers. I will look into it.

@Algorithm5838 you right

image


here is the pattern for a test

```
$1$ $1$
$1 $2 $3 $3
`$1` $2 $3 `$4`
$x$ $hello world$  $x$  $1 $2 $3 $4
$x$ $hello world$  $x$  $1 $2 $3 $4 $x$ $hello world$  $x$
$1 $2 $3 $4 $x$
```
```
$1 2 $3 $4 $5
```

Comment (Algorithm5838) [reply app/components/markdown.tsx]: Fixed.

**Pull request closed without merging**